gadget/layout.go

```diff
@@ -0,0 +1,214 @@
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package gadget
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+type LayoutConstraints struct {
+	NonMBRStartOffset Size
+	SectorSize        Size
+}
+
+type PositionedVolume struct {
+	// Size is the total size of the volume
+	Size Size
+	// Structures are sorted in order of 'appearance' in the volume
+	Structures []PositionedStructure
+}
+
+// PositionedStructure describes a VolumeStructure that has been positioned
+// within the volume
+type PositionedStructure struct {
+	*VolumeStructure
+	// StartOffset defines the start offset of the structure within the
+	// enclosing volume
+	StartOffset Size
+	// Index of the structure definition in gadget YAML
+	Index int
+
+	// PositionedContent is a list of raw content included in this structure
+	PositionedContent []PositionedContent
+}
+
+func (p PositionedStructure) String() string {
+	return fmtIndexAndName(p.Index, p.Name)
+}
+
+type byStartOffset []PositionedStructure
+
+func (b byStartOffset) Len() int           { return len(b) }
+func (b byStartOffset) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+func (b byStartOffset) Less(i, j int) bool { return b[i].StartOffset < b[j].StartOffset }
+
+// PositionedContent describes raw content that has been positioned within the
+// encompassing structure
+type PositionedContent struct {
+	*VolumeContent
+
+	// StartOffset defines the start offset of this content image
+	StartOffset Size
+	// Size is the maximum size occupied by this image
+	Size Size
+}
+
+// LayOutVolume attempts to lay out the volume using constraints and returns a
+// fully positioned description of the resulting volume
+func LayOutVolume(gadgetRootDir string, volume *Volume, constraints LayoutConstraints) (*PositionedVolume, error) {
+	previousEnd := Size(0)
+	farthestEnd := Size(0)
+	structures := make([]PositionedStructure, len(volume.Structure))
+
+	if constraints.SectorSize == 0 {
+		return nil, fmt.Errorf("cannot lay out volume, invalid constraints: sector size cannot be 0")
+	}
+
+	for idx, s := range volume.Structure {
+		var start Size
+		if s.Offset == nil {
+			if previousEnd < constraints.NonMBRStartOffset {
+				start = constraints.NonMBRStartOffset
+			} else {
+				start = previousEnd
+			}
+		} else {
+			start = *s.Offset
+		}
+
+		end := start + s.Size
+		ps := PositionedStructure{
+			VolumeStructure: &volume.Structure[idx],
+			StartOffset:     start,
+			Index:           idx,
+		}
+
+		if ps.EffectiveRole() != "mbr" {
```

Comment:
Back to my earlier comment, perhaps mbr (gpt..) could indeed be a constant instead?